### DISCUSSION TITLE
Why is the use of ActiveJob::Base enforced?

### DISCUSSION THREAD

**Discussion by edjames** [RailsEventStore/rails_event_store#1219] [Q&A] [answered]
Hello.

In `RailsEventStore::ActiveJobScheduler#verify` there is a hard-coded check to ensure that the subscriber inherits from `ActiveJob::Base`. Is there a reason for this?

With 2 small monkey patches I can very easily disable this check and safely use workers native Sidekiq workers. I'm wondering if there is a reason for this or whether I can raise a small pull request to make this check optional.

Here is my code:

<details>
 <summary>Monkey patch for RES</summary>

```
module RailsEventStore
  class ActiveJobScheduler
    def verify(subscriber)
      # Class === subscriber && !!(subscriber < ActiveJob::Base) ==> comment out original
      Class === subscriber
    end
  end
end
```
</details>

<details>
 <summary>ApplicationJob without ActiveJob::Base</summary>

```
class ApplicationJob # < ActiveJob::Base ==> do not inherit
  include Sidekiq::Worker

  # simple method proxy
  def self.perform_later(*args)
    perform_async(*args)
  end

  def event_store
    Rails.configuration.event_store
  end

end
```
</details>

The reason I want to use the native Sidekiq workers is because of the much better control around retry logic and job visibility in the Sidekiq web UI. It also means I can have consistency between jobs used in my RES implementation as well as general jobs.

Thanks.

**Comment by mostlyobvious** [comment, marked as answer]
> Is there a reason for this?

It's an ActiveJob scheduler. It's assuming you always have ActiveJob framework in Rails, when installing `rails_event_store` that depends on Rails. It does not assume what's below ActiveJob facade, that's the whole point of the facade.

Whether you have Sidekiq or not in your project, and whether you want to use it via ActiveJob or directly is up to you. There's nothing in `rails_event_store` preventing you from doing this. Quite the contrary.

> With 2 small monkey patches

There's no need to monkey patch at all. You'll want to pass a different scheduler that talks to Sidekiq directly, skipping ActiveJob.

There's even 20-LOC-gem released to make it easier:
https://github.com/RailsEventStore/rails_event_store/blob/master/contrib/ruby_event_store-sidekiq_scheduler/lib/ruby_event_store/sidekiq_scheduler.rb#L3-L20

It has a README describing how to pass this dependency:
```
event_store = RailsEventStore::Client.new(
  dispatcher: RailsEventStore::AfterCommitAsyncDispatcher.new(scheduler: RubyEventStore::SidekiqScheduler.new(serializer: YAML),
)
```
That scheduler already switches `perform_later` for `perform_async` and there's no need for method proxy on the handler side.


PS. If that's missing in the docs or was hard to find, feel free to change it. I'm going to move this issue to discussions section.

**Reply by edjames** [reply]
Thanks very much! That's exactly what I was looking for. When I have time I will update the docs as I had no idea there was a `contrib` directory with these useful things.

**Comment by mostlyobvious** [comment]
Just in case — here are the default choices when going with `RailsEventStore::Client.new`:

https://github.com/RailsEventStore/rails_event_store/blob/master/rails_event_store/lib/rails_event_store/client.rb#L10-L12
